### PR TITLE
Flinks

### PR DESCRIPTION
**Pull request by xuerongCode**
Add StringParser(), DoubleStringParser in AnyParser,
modify UnknownParser
add UnknownFObjectArrayParser
add outputJSON mode

### CODE REVIEW

**Comment by kirkeaton** [comment]
Why the need for a StringArrayParser specifically. Was ArrayParser on its own not working?

File: src/foam/lib/json/OutputterMode.js [hunk 1]
```diff
@@ -22,6 +22,10 @@ foam.ENUM({
     {
       name: 'FULL',
       label: 'Full'
+    },
+    {
+      name: 'SIMPLE',
```
Comment (kgrgreer) [inline src/foam/lib/json/OutputterMode.js]: Document what this means.

File: src/foam/lib/json/StringDoubleArrayParser.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/StringDoubleArrayParser.java]: 2018

File: src/foam/lib/json/UnknownArrayParser.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownArrayParser.java]: 2018

File: src/foam/lib/json/UnknownFObjectArray.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownFObjectArray.js]: 2018

File: src/foam/lib/json/UnknownKeyValueParser0.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/**
+  * @license
+  * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownKeyValueParser0.java]: 2018

File: src/foam/lib/json/UnknownObjectParser.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/**
+  * @license
+  * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownObjectParser.java]: 2018

File: src/foam/lib/json/UnknownParser.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownParser.java]: 2018

File: src/foam/lib/json/UnknownReferenceParser.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownReferenceParser.java]: 2018

File: src/foam/lib/json/UnknownStringParser.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/lib/json/UnknownStringParser.java]: 2018

File: src/foam/java/PropertyInfo.js [hunk 10]
```diff
@@ -31,6 +31,11 @@ foam.CLASS({
       class: 'Boolean',
       name: 'storageTransient'
     },
+    {
+      class: 'Boolean',
+      name: 'showDefault',
```
Comment (kgrgreer) [inline src/foam/java/PropertyInfo.js]: Rename to 'outputDefaultValues' to be the same as JS.

File: src/foam/java/PropertyInfo.js [hunk 11]
```diff
@@ -31,6 +31,11 @@ foam.CLASS({
       class: 'Boolean',
       name: 'storageTransient'
     },
+    {
+      class: 'Boolean',
```
Comment (kgrgreer) [inline src/foam/java/PropertyInfo.js]: Why is this in PropertyInfo?

File: src/foam/lib/json/Outputter.java [hunk 12]
```diff
@@ -32,6 +32,7 @@ protected SimpleDateFormat initialValue() {
   protected StringWriter  stringWriter_ = null;
   protected PrintWriter   writer_;
   protected OutputterMode mode_;
+  protected boolean       showDefault_ = false;
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: Rename to outputDefaultValues to be compatible is JS.

File: src/foam/lib/json/Outputter.java [hunk 13]
```diff
@@ -211,7 +211,7 @@ protected void outputFObject(FObject o) {
       PropertyInfo prop = (PropertyInfo) i.next();
       if ( mode_ == OutputterMode.NETWORK && prop.getNetworkTransient() ) continue;
       if ( mode_ == OutputterMode.STORAGE && prop.getStorageTransient() ) continue;
-      if ( showDefault_ == false && ! prop.isSet(o) ) continue;
+      if ( outputDefaultValues_ == false && ! prop.isSet(o) ) continue;
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: ! is shorter than == false

**Pull request merged**